### PR TITLE
GrantZ

### PR DESCRIPTION
**Pull request by mbeacom**
##### Description

[BALMER!](http://www.youtube.com/watch?v=HMhrRovP9qA "Developers, Developers, Developers")

##### Checklist

- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://docs.gitcoin.co/mk_contributors/#step-4-commit)

### CODE REVIEW

File: app/grants/views.py [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+            'grant': grant,
+            'text': _('You already have an active subscription for this grant.')
+        }
+        return TemplateResponse(request, 'grants/shared/error.html', params)
+
+    # make sure a user can only create one subscription per grant
+    if request.method == 'POST':
+        subscription = Subscription()
+
+        subscription.subscription_hash = request.POST.get('subscription_hash', '')
```
Comment (owocki) [inline app/grants/views.py]: is it worth doing some validation on the POST?

File: app/grants/views.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+
+    profile = request.user.profile if request.user.is_authenticated and request.user.profile else None
+    _grants = Grant.objects.prefetch_related('team_members') \
+        .filter(Q(admin_profile=profile) | Q(team_members__in=[profile])).order_by(sort)
+    sub_grants = Grant.objects.filter(subscriptions__contributor_profile=profile).order_by(sort)
+
+    paginator = Paginator(_grants, limit)
+    grants = paginator.get_page(page)
+
+    history = [{
```
Comment (owocki) [inline app/grants/views.py]: this is hard coded / needs to be driven by the backend

File: app/marketing/mails.py [hunk 3]
```diff
@@ -105,6 +107,96 @@ def send_mail(from_email, _to_email, subject, body, html=False,
...
+        setup_lang(to_email)
+        html, text, subject = render_grant_cancellation_email(grant, subscription)
+
+        if not should_suppress_notification_email(to_email, 'grant_cancellation'):
+            send_mail(from_email, to_email, subject, text, html, categories=['transactional', func_name()])
+    finally:
+        translation.activate(cur_language)
+
+
+def subscription_terminated(grant, subscription):
```
Comment (owocki) [inline app/marketing/mails.py]: which of these emails still need to be WIRED up still?  worth throwing a TODO in here

File: app/retail/emails.py [hunk 4]
```diff
@@ -67,6 +68,102 @@ def premailer_transform(html):
     return p.transform()
 
 
+def render_new_grant_email(grant):
+    params = {'grant': grant}
+    response_html = premailer_transform(render_to_string("emails/grants/new_grant.html", params))
+    response_txt = render_to_string("emails/grants/new_grant.txt", params)
+    subject = "Your Gitcoin Grant"
```
Comment (owocki) [inline app/retail/emails.py]: IMHO its worth putting the grant title in each of these subject lines (people may have multiple grants)

File: app/retail/emails.py [hunk 5]
```diff
@@ -602,7 +699,7 @@ def render_new_bounty_roundup(to_email):
 <h3>What else is new?</h3>
     <ul>
         <li>
-            Colony is running an on-going bug bounty program using Gitcoin for the <a href="https://github.com/JoinColony/colonyNetwork">colonyNetwork repository</a>. 
+            Colony is running an on-going bug bounty program using Gitcoin for the <a href="https://github.com/JoinColony/colonyNetwork">colonyNetwork repository</a>.
```
[the remark this answers is not in the document]
Comment (owocki) [reply app/retail/emails.py]: not sure why this was changed

File: app/retail/templates/shared/footer.html [hunk 6]
```diff
@@ -61,7 +61,6 @@ <h5>{% trans "Product" %}</h5>
       <a href="{% url "explorer" %}">{% trans "Issue Explorer" %}</a>
       <a href="{% url "tools" %}">{% trans "Tools" %}</a>
       <a href="{% url "kudos_main" %}">{% trans "Kudos" %}</a>
-      <a href="{% url "grants" %}">{% trans "Grants" %}</a>
```
Comment (owocki) [inline app/retail/templates/shared/footer.html]: why was this removed

File: app/retail/templates/emails/grants/transaction_summary.html [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+  You should have received a copy of the GNU Affero General Public License
+  along with this program. If not, see <http://www.gnu.org/licenses/>.
+{% endcomment %}
+{% load i18n static %}
+
+{% block content %}
+
+<style>
+  hr {
+    width: 80%;
```
Comment (owocki) [inline app/retail/templates/emails/grants/transaction_summary.html]: should some of these be in a shared template?  they look repeated

File: app/retail/templates/emails/grants/thank_you_for_supporting.html [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+  You should have received a copy of the GNU Affero General Public License
+  along with this program. If not, see <http://www.gnu.org/licenses/>.
+{% endcomment %}
+{% load i18n static grants_extra %}
+
+{% block content %}
+
+<style>
+  hr {
```
Comment (owocki) [inline app/retail/templates/emails/grants/thank_you_for_supporting.html]: DRY

File: app/retail/templates/emails/grants/subscription_terminated.html [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+<hr>
+<div class="grantInfo">
+  <a href="{% url 'grants:details' grant.id grant.slug %}">
+    <img src="{{ grant.logo.url }}" style="width:50px; height: 50px;"/>
+  </a>
+  <br>
+  <a class="grantInfoName" href="{% url 'grants:details' grant.id grant.slug %}">{{ grant.title }}</a>
+  <p>{{ grant.description }}</p>
+  <div class="progress">
+    <!-- #TODO - Revisit which object to use to pull grant status. -->
```
Comment (owocki) [inline app/retail/templates/emails/grants/subscription_terminated.html]: can this be done now?  we should have the status off the DB ojects

File: app/retail/templates/emails/grants/support_cancellation.html [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+  You should have received a copy of the GNU Affero General Public License
+  along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+{% endcomment %}
+{% load i18n static %}
+
+{% block content %}
+
+<style>
+  hr {
```
Comment (owocki) [inline app/retail/templates/emails/grants/support_cancellation.html]: DRY

File: app/assets/v2/js/grants/detail.js [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    editableFields.forEach(field => disableEdit(field));
+    copyDuplicateDetails();
+  });
+
+  $('#cancel-details').on('click', (event) => {
+    $('#edit-details').removeClass('hidden');
+    $('#save-details').addClass('hidden');
+    $('#cancel-details').addClass('hidden');
+
+    // TODO: Reset value
```
[the remark this answers is not in the document]
Comment (thelostone-mc) [reply app/assets/v2/js/grants/detail.js]: oh that is done :P
forgot to remove the todo

File: app/retail/emails.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/retail/emails.py]: Looks like the only change was an extra space was removed at the end of the sentence. The line difference is adding more code above this line.

File: app/retail/emails.py [hunk 12]
```diff
@@ -611,7 +708,7 @@ def render_new_bounty_roundup(to_email):
     </ul>
 </p>
 <p>
-Thanks for reading to the end!  <a href="https://gitcoin.co/kudos/redeem/red_staplers_for_all">Here's a kudos just for you :)</a>. 
+Thanks for reading to the end!  <a href="https://gitcoin.co/kudos/redeem/red_staplers_for_all">Here's a kudos just for you :)</a>.
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/retail/emails.py]: Same for this one. Trailing space removed.

File: app/retail/emails.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (thelostone-mc) [reply app/retail/emails.py]: To remove the trailing white spacing

File: app/assets/v2/js/grants/detail.js [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    editableFields.forEach(field => disableEdit(field));
+  });
+
+  $('#cancel_grant').on('click', function(e) {
+    var img_src = static_url + 'v2/images/grants/cancel-grants-icon.png';
+    var content = $.parseHTML(
+      '<div>' +
```
[the remark this answers is not in the document]
Comment (thelostone-mc) [reply app/assets/v2/js/grants/detail.js]: @SaptakS

File: app/assets/v2/js/grants/detail.js [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+      modalClass: 'modal cancel_grants'
+    });
+
+    $('.modal-cancel-grants').on('click', function(e) {
+      let contract_address = $('#contract_address').val();
+      let deployedSubscription = new web3.eth.Contract(compiledSubscription.abi, contract_address);
+
+      web3.eth.getAccounts(function(err, accounts) {
+        deployedSubscription.methods.endContract()
+          .send({from: accounts[0], gasPrice: 4000000000})
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/assets/v2/js/grants/detail.js]: The `gasPrice` is hardcoded only because we've had bigger fish to fry. I will add it to my list to implement.

File: app/assets/v2/js/grants/fund.js [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+        if (periodSeconds) {
+          realPeriodSeconds = periodSeconds;
+        }
+      }
+
+      let deployedSubscription = new web3.eth.Contract(compiledSubscription.abi, data.contract_address);
+      let deployedToken;
+      let selected_token;
+
+      if (data.token_address != '0x0000000000000000000000000000000000000000') {
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/assets/v2/js/grants/fund.js]: Correct. However, `'0x0000000000000000000000000000000000000000'` is the 0 value for a contract and in `grants` is used to represent `Any Token` since user selects no contract when chosing this option in `grants/new`.

File: app/assets/v2/js/grants/fund.js [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+                    alert('Your subscription data failed to save. Please try again.');
+
+                  });
+                });
+              });
+            });
+          }).on('confirmation', function(confirmationNumber, receipt) {
+            console.log('receipt', receipt);
+            //
+            // $.each($(form).serializeArray(), function() {
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/assets/v2/js/grants/fund.js]: Not sure why this is commented out...

File: app/assets/v2/js/grants/new.js [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+                  setTimeout(function() {
+                    callFunctionWhenTransactionMined(transactionHash);
+                  }, 1000);
+                }
+              });
+            };
+
+            callFunctionWhenTransactionMined(transactionHash);
+          });
+          // .on('receipt', function(receipt) {
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/assets/v2/js/grants/new.js]: @owocki This is the code the should be reimplemented when the intermittent contract deploy issue is fixed with web3.js. It was left commented out for this reason instead of removing. Should we remove it?

File: app/assets/v2/js/grants/shared.js [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+  $('.grant_waiting').show();
+  waitingStateActive();
+};
+
+var waitingStateActive = function() {
+  $('.bg-container').show();
+  $('.loading_img').addClass('waiting-state ');
+  $('.waiting_room_entertainment').show();
+  $('.issue-url').html('<a href="' + document.issueURL + '">' + document.issueURL + '</a>');
+  var secondsBetweenQuoteChanges = 30;
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/assets/v2/js/grants/shared.js]: The general/shared.js throws an error if tagged. But will defer to @thelostone-mc @SaptakS

File: app/dashboard/models.py [hunk 19]
```diff
@@ -1516,6 +1516,14 @@ class Activity(models.Model):
         ('bounty_removed_by_staff', 'Removed from Bounty by Staff'),
         ('bounty_removed_by_funder', 'Removed from Bounty by Funder'),
         ('new_crowdfund', 'New Crowdfund Contribution'),
+        # Grants
+        ('new_grant', 'New Grant'),
+        ('update_grant', 'Updated Grant'),
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/dashboard/models.py]: @owocki I dont think they've been implemented. @mark has been wanting to get them in there, but I dont fully understand what the goal is with the activities model. And we've havent had time to go over it yet.

File: app/grants/templates/grants/milestones.html [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+    </div>
+
+    {% include 'shared/bottom_notification.html' %}
+    {% include 'shared/analytics.html' %}
+    {% include 'shared/footer_scripts.html' %}
+    {% include 'shared/footer.html' %}
+  </body>
+
+  <!-- jQuery -->
+  <script src="{% static "v2/js/abi.js" %}"></script>
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/grants/templates/grants/milestones.html]: We dont need most of the script tags on this page. Only need `milestones.js`.

File: app/grants/templates/grants/shared/shared_scripts.html [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+{% endcomment %}
+{% load static %}
+
+<script src="{% static "v2/js/web3.min.js" %}"></script>
+<script>
+  var web3 = new Web3(Web3.givenProvider || "ws://localhost:8546");
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/grants/templates/grants/shared/shared_scripts.html]: This is what the documentation says to have. Not sure that we need it though.

File: app/grants/templates/grants/shared/waiting_state.html [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+  You should have received a copy of the GNU Affero General Public License
+  along with this program. If not, see <http://www.gnu.org/licenses/>.
+{% endcomment %}
+{% load static i18n %}
+<div class="row no-gutter grant_waiting" style="display: none;">
+  <div class="col-12 offset-md-1 col-md-10 offset-lg-1 col-lg-8 body ">
+    {% include 'shared/bounty_waiting_state.html' %}
+    <div class="row ml-2 mr-2 pl-4 pr-4 waiting_room_entertainment" style="display: none; text-align: center">
+      {% include 'shared/rain.html' with class="color" %}
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/grants/templates/grants/shared/waiting_state.html]: @thelostone-mc @SaptakS

File: app/retail/templates/emails/grants/transaction_summary.html (hunk 7 above)
Comment (thelostone-mc) [reply app/retail/templates/emails/grants/transaction_summary.html]: You mean move the css into standalone template file ? 🤔 
as in it does seem a lil weird but we could do that

File: app/grants/views.py [hunk 23]
```diff
@@ -0,0 +1,430 @@
...
+        'card_desc': _('Provide sustainable funding for Open Source with Gitcoin Grants'),
+        'grants': grants,
+        'grants_count': _grants.count(),
+        'keywords': get_keywords(),
+    }
+    return TemplateResponse(request, 'grants/index.html', params)
+
+
+@csrf_exempt
+def grant_details(request, grant_id, grant_slug):
```
[the remark this answers is not in the document]
Comment (captnseagraves) [reply app/grants/views.py]: Yeah, we probably should be.

File: app/grants/views.py (hunk 1 above)
Comment (captnseagraves) [reply app/grants/views.py]: Yes, probably.

File: app/grants/views.py (hunk 2 above)
Comment (captnseagraves) [reply app/grants/views.py]: Waiting on the miner

File: app/retail/templates/emails/grants/transaction_summary.html (hunk 7 above)
Comment (captnseagraves) [reply app/retail/templates/emails/grants/transaction_summary.html]: Yes, could be in a shared template. Will put refactor onto list.

File: app/retail/templates/emails/grants/subscription_terminated.html (hunk 9 above)
Comment (captnseagraves) [reply app/retail/templates/emails/grants/subscription_terminated.html]: Yes, I think this comment has been addressed and is a holdover from Anish's original contribution.

File: app/assets/v2/js/grants/detail.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (SaptakS) [reply app/assets/v2/js/grants/detail.js]: @thelostone-mc @owocki currently it is being used like this in many different places in the platform for the modal. So I made it like this to keep continuity and so I could refactor all in the same way when doing. But if you strongly feel I should convert this to x-js, I can do that.
